Gitweb:     http://git.kernel.org/git/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commit;h=8246934b7cf99d1f0c053d57890775e5d0df9c33
Commit:     8246934b7cf99d1f0c053d57890775e5d0df9c33
Parent:     c351f2dd542a3980e96cf128e06d19f784c5ea3e
Author:     Lin Ming <ming.m.lin@intel.com>
AuthorDate: Thu Apr 10 19:06:41 2008 +0400
Committer:  Len Brown <len.brown@intel.com>
CommitDate: Tue Apr 22 14:29:29 2008 -0400

    ACPICA: Fix for SizeOf when used with Buffers and Packages
    
    Fixed a problem with the SizeOf operator when used with Package
    and Buffer objects. These objects have deferred execution for some
    arguments, and the execution is now completed before the SizeOf is
    executed. This problem caused unexpected AE_PACKAGE_LIMIT errors
    on some systems.
    
    http://bugzilla.kernel.org/show_bug.cgi?id=9558
    
    Signed-off-by: Lin Ming <ming.m.lin@intel.com>
    Signed-off-by: Bob Moore <robert.moore@intel.com>
    Signed-off-by: Alexey Starikovskiy <astarikovskiy@suse.de>
    Signed-off-by: Len Brown <len.brown@intel.com>
---
 drivers/acpi/executer/exoparg1.c |   22 +++++++++++++++++-----
 1 files changed, 17 insertions(+), 5 deletions(-)

Index: 2008-devel/drivers/acpi/executer/exoparg1.c
===================================================================
--- 2008-devel.orig/drivers/acpi/executer/exoparg1.c
+++ 2008-devel/drivers/acpi/executer/exoparg1.c
@@ -739,26 +739,38 @@ acpi_status acpi_ex_opcode_1A_0T_1R(stru
 			value = acpi_gbl_integer_byte_width;
 			break;
 
-		case ACPI_TYPE_BUFFER:
-			value = temp_desc->buffer.length;
-			break;
-
 		case ACPI_TYPE_STRING:
 			value = temp_desc->string.length;
 			break;
 
+		case ACPI_TYPE_BUFFER:
+
+			/* Buffer arguments may not be evaluated at this point */
+
+			status = acpi_ds_get_buffer_arguments(temp_desc);
+			value = temp_desc->buffer.length;
+			break;
+
 		case ACPI_TYPE_PACKAGE:
+
+			/* Package arguments may not be evaluated at this point */
+
+			status = acpi_ds_get_package_arguments(temp_desc);
 			value = temp_desc->package.count;
 			break;
 
 		default:
 			ACPI_ERROR((AE_INFO,
-				    "Operand is not Buf/Int/Str/Pkg - found type %s",
+				    "Operand must be Buffer/Integer/String/Package - found type %s",
 				    acpi_ut_get_type_name(type)));
 			status = AE_AML_OPERAND_TYPE;
 			goto cleanup;
 		}
 
+		if (ACPI_FAILURE(status)) {
+			goto cleanup;
+		}
+
 		/*
 		 * Now that we have the size of the object, create a result
 		 * object to hold the value
